src/services/sync-component-with-file/config-file-changes-transformer.ts

```diff
@@ -0,0 +1,110 @@
+import { ComponentChanges, ComponentSyncPayload, ComponentUnlinkPayload, ModifiedFilePayload } from '../../types';
+import { listFeatures } from '../feature-flags';
+
+const isApplicableForBothIdentifiersOrOnlyForId = (isApplicableForId: boolean) => {
+  const { isCreateFromYamlEnabled } = listFeatures();
+
+  return isCreateFromYamlEnabled || isApplicableForId;
+};
+
+const isFileMovedDetectionEnabledForName = (isApplicableForName: boolean) => {
+  const { isCreateFromYamlEnabled } = listFeatures();
+
+  return isCreateFromYamlEnabled && isApplicableForName;
+};
+
+const isFileIdentifierChanged = (oldFile: ComponentUnlinkPayload, newFile: ComponentSyncPayload): boolean => {
+  const { isCreateFromYamlEnabled } = listFeatures();
+  const { id: oldId, name: oldName } = oldFile.componentYaml;
+  const { id: newId, name: newName } = newFile.componentYaml;
+
+  const isIdChanged = oldId !== newId;
+  const isNameFileIdentifier = oldId === undefined && newId === undefined;
+  const isIdAdded = oldId === undefined && newId !== undefined;
+
+  const isFileMovedAsGitlabRename = newFile.filePath !== newFile.previousFilePath;
+  const isNameChangedWithFileMove = isNameFileIdentifier && isFileMovedAsGitlabRename && oldName !== newName;
+  const isIdFileIdentifierChanged = isIdChanged && !isIdAdded;
+
+  return isCreateFromYamlEnabled ? isNameChangedWithFileMove || isIdFileIdentifierChanged : isIdChanged;
+};
+
+export const detectMovedFilesAndUpdateComponentChanges = (changes: ComponentChanges): ComponentChanges => {
+  let creates = [...changes.componentsToCreate];
```

Comment:
this function breaks the design principle outlined in magnets [dev guide](https://hello.atlassian.net/wiki/spaces/ServiceMatrix/pages/1741394111/Magnets+developers+coding+guidelines#Code-Design-Principles) and SOLID design principle. We need to look into simplifying in follow up PR, as debugging any bugs in this part of the code will be extremely difficult. For now, do we have any documentation where i can read the approach taken to determine if a file was moved?

Reply:
Yep, I definitely agree that this function could be hard to debug and would be open to spending some time doing a refactor of it to try to simplify it. I think we have similar code committed to the github app, so spending some time cleaning it up and standardizing it would be worthwhile - I noticed some small differences between them that I think are differences in https://softwareteams.atlassian.net/browse/COMPASS-13170 being taken into account for this app.
As for how the approach for if the file was moved or not, generally, we should be able to track all changes except if the `name` was changed in the file in the same push that it was moved in the repo. Ill can add an explict section to the project doc talking about this.